logstash-core/lib/logstash/util/wrapped_synchronous_queue.rb

```diff
@@ -37,5 +37,231 @@ def take
     def poll(millis)
       @queue.poll(millis, TimeUnit::MILLISECONDS)
     end
+
+    def write_client
+      WriteClient.new(self)
+    end
+
+    def read_client
+      ReadClient.new(self)
+    end
+
+    class ReadClient
+      # We generally only want one thread at a time able to access pop/take/poll operations
+      # from this queue. We also depend on this to be able to block consumers while we snapshot
+      # in-flight buffers
+
+      def initialize(queue)
+        @queue = queue
+        @mutex = Mutex.new
+        @inflight_batches = {}
```

Comment:
ok, good point. I would add a comment to that effect here somewhere and also in the Pipeline where the `read_client` is called so that if someone figures we should have multiple read clients this current limitation will be obvious